tracer/src/Datadog.Trace/AppSec/Security.cs

```diff
@@ -243,30 +216,152 @@ private static void AddAppsecSpecificInstrumentations()
+        }
+
+        private void UpdateStatus()
+        {
+            if (_enabled == _settings.Enabled) { return; }
+            lock (_settings)
+            {
+                if (_settings.Enabled)
+                {
+                    if (_waf != null)
+                    {
+                        _waf.Dispose();
+                    }
+
+                    _waf = Waf.Waf.Create(_settings.ObfuscationParameterKeyRegex, _settings.ObfuscationParameterValueRegex, _settings.Rules);
+                    if (_waf?.InitializedSuccessfully ?? false)
+                    {
+                        EnableWaf();
+                    }
+                    else
+                    {
+                        _settings.Enabled = false;
+                    }
+                }
+
+                if (!_settings.Enabled)
+                {
+                    DisableWaf();
+                }
+            }
+        }
+
+        private void EnableWaf()
+        {
+            if (!_enabled)
+            {
+                Log.Information("AppSec Enabled");
+                _instrumentationGateway.EndRequest += RunWaf;
+                _instrumentationGateway.PathParamsAvailable += AggregateAddressesInContext;
+                _instrumentationGateway.BodyAvailable += AggregateAddressesInContext;
+#if NETFRAMEWORK
+                if (_usingIntegratedPipeline == null)
+                {
+                    try
+                    {
+                        _usingIntegratedPipeline = TryGetUsingIntegratedPipelineBool();
+                    }
+                    catch (Exception ex)
+                    {
+                        _usingIntegratedPipeline = false;
+                        Log.Error(ex, "Unable to query the IIS pipeline. Request and response information may be limited.");
+                    }
+                }
+
+                if (_usingIntegratedPipeline.Value)
+                {
+                    _instrumentationGateway.LastChanceToWriteTags += InstrumentationGateway_AddHeadersResponseTags;
+                }
+
+#else
+                _instrumentationGateway.LastChanceToWriteTags += InstrumentationGateway_AddHeadersResponseTags;
+#endif
+                AddAppsecSpecificInstrumentations();
+
+                _instrumentationGateway.EndRequest += ReportWafInitInfoOnce;
+                _rateLimiter = _rateLimiter ?? new AppSecRateLimiter(_settings.TraceRateLimit);
+
+                _enabled = true;
+            }
+        }
+
+        private void DisableWaf()
+        {
+            if (_enabled)
+            {
+                Log.Information("AppSec Disabled");
+
+                _instrumentationGateway.EndRequest -= RunWaf;
+                _instrumentationGateway.PathParamsAvailable -= AggregateAddressesInContext;
+                _instrumentationGateway.BodyAvailable -= AggregateAddressesInContext;
+                _instrumentationGateway.LastChanceToWriteTags -= InstrumentationGateway_AddHeadersResponseTags;
```

Comment:
This isn't always set on .NET Framework, so you might need to add some field to keep track of whether it was added or not

Reply:
Actually, unsubscribing an event multiple times appears to be fine based on some Google searching. TIL

Reply:
Yes, that was the reason why I left it this way, for simplicity.